src/cls/rgw/cls_rgw_types.h

```diff
@@ -718,39 +719,33 @@ struct rgw_bucket_category_stats {
 };
 WRITE_CLASS_ENCODER(rgw_bucket_category_stats)
 
-enum class cls_rgw_reshard_status : uint8_t {
-  NOT_RESHARDING  = 0,
-  IN_PROGRESS     = 1,
-  DONE            = 2
-};
-
-inline std::string to_string(const cls_rgw_reshard_status status)
+inline std::string to_string(const rgw::BucketReshardState status)
 {
   switch (status) {
-  case cls_rgw_reshard_status::NOT_RESHARDING:
+  case rgw::BucketReshardState::NOT_RESHARDING:
     return "not-resharding";
     break;
-  case cls_rgw_reshard_status::IN_PROGRESS:
+  case rgw::BucketReshardState::IN_PROGRESS:
     return "in-progress";
     break;
-  case cls_rgw_reshard_status::DONE:
+  case rgw::BucketReshardState::DONE:
     return "done";
     break;
   };
   return "Unknown reshard status";
 }
 
 struct cls_rgw_bucket_instance_entry {
-  using RESHARD_STATUS = cls_rgw_reshard_status;
+  using RESHARD_STATUS = rgw::BucketReshardState;
   
-  cls_rgw_reshard_status reshard_status{RESHARD_STATUS::NOT_RESHARDING};
-  std::string new_bucket_instance_id;
+  rgw::BucketReshardState reshard_status{RESHARD_STATUS::NOT_RESHARDING};
+  std::string bucket_instance_id;
```

Comment:
as i understand it, `struct cls_rgw_bucket_instance_entry` is just for implementing `RGWRados::Bucket::UpdateIndex::guard_reshard()`. its responsibilities are:
* prevent writes to the index during reshard by adding `cls_rgw_guard_bucket_resharding()` to the bucket index op,
* sleep and poll the reshard state while it's 'in-progress', and
* when reshard state is 'done', transition to the new bucket index shard and retry the index op there

the `new_bucket_instance_id` and `num_shards` fields appear to be entirely unused, so i think they're safe to remove as long as we fix up the `decode()`

the cls call for `cls_rgw_guard_bucket_resharding()` just returns an error when cls_rgw_bucket_instance_entry.reshard_status is in-progress or done

when `RGWRados::block_while_resharding()` polls on the `cls_rgw_get_bucket_resharding()` status, it needs to differentiate between 'in-progress' (where it should sleep and poll again) and 'done' (where it should transition to a new bucket index shard)

once it reaches 'done', the `fetch_new_bucket_id()` function should instead fetch the latest `current_index` layout from the bucket instance, select a new `BucketShard` based on that layout, and resend the op to that index shard

Reply:
@cbodley to test the current guard_reshard()'s functionality, running any operation that updates bucket index should suffice, right?

Reply:
yeah, uploading an object is probably the easiest way to trigger it

Reply:
@cbodley modified. could you take a look please?